src/OtcNexus.sol

```diff
@@ -0,0 +1,604 @@
+    }
+
+    /**
+     * @notice take a fixed price RFS
+     * @param _id id of the RFS
+     * @param _paymentTokenAmount amount of tokens used to pay for the RFS
+     * @param index index of the token used to buy
+     * @param expectedRfsType type of the Rfs: FIXED or DYNAMIC
+     * @return success
+     */
+    function takeRfs(
+        uint256 _id,
+        uint256 _paymentTokenAmount,
+        uint256 index,
+        RfsType expectedRfsType
+    ) internal returns (bool success) {
+        RFS memory rfs = getRfs(_id);
+
+        // sanity checks
+        if (rfs.typeRfs != expectedRfsType) revert OtcNexus__InvalidRfsType();
+        if (rfs.removed) revert OtcNexus__RfsRemoved();
+        if (_paymentTokenAmount == 0) revert OtcNexus__InvalidTokenAmount();
+        if (index >= rfs.tokensAccepted.length) revert OtcNexus__InvalidTokenIndex();
+        if (
+            IERC20(rfs.tokensAccepted[index]).allowance(msg.sender, address(this)) <
+            _paymentTokenAmount
+        ) revert OtcNexus__AllowanceToken1TooLow();
+
+        //0.25% fee on the payment token of the taker FLAT FEE
+        (uint256 feeAmount1, uint256 _paymentTokenAmountAfterTakerFee) = takeFee(
+            takerFee,
+            _paymentTokenAmount
+        );
+        // compute _amount0 (depends on rfs.typeRfs, so this is left to takeFixedRfs/takeDynamicRfs)
+        uint256 _amountBuying = computeAmount0(
+            _id,
+            _paymentTokenAmountAfterTakerFee,
+            index,
+            expectedRfsType
+        );
+        if (_amountBuying > rfs.amount0) revert OtcNexus__Amount0TooHigh();
+
+        //0.25% fee on the token0 of the maker FLAT FEE if DEPOSITED, otherwise 0.5% if APPROVED
+        (uint256 feeAmount2, uint256 _paymentTokenAmountAfterMakerFee) = takeFee(
+            rfs.interactionType == TokenInteractionType.TOKEN_DEPOSITED
+                ? makerFeeIfDeposited
+                : makerFeeIfNotDeposited,
+            _paymentTokenAmountAfterTakerFee
+        );
+        // transfer token1 to the maker
+        success = IERC20(rfs.tokensAccepted[index]).transferFrom(
+            msg.sender,
+            rfs.maker,
+            _paymentTokenAmountAfterMakerFee
+        );
+        if (!success) revert OtcNexus__TransferToken1Failed();
+
+        success = IERC20(rfs.tokensAccepted[index]).transferFrom(
+            msg.sender,
+            address(this),
+            feeAmount1 + feeAmount2
+        );
+        if (!success) revert OtcNexus__TransferToken1Failed();
+
+        // transfer token0 to the taker
+        if (rfs.interactionType == TokenInteractionType.TOKEN_DEPOSITED) {
+            success = IERC20(rfs.token0).transfer(msg.sender, _amountBuying);
+        } else {
+            success = IERC20(rfs.token0).transferFrom(rfs.maker, msg.sender, _amountBuying);
+        }
+
+        if (!success) revert OtcNexus__TransferToken0Failed();
+
+        // update RFS
```

Comment:
@Lojesad#7078 
There is also one problem:
        // update RFS
        updateRfs(rfs, _amountBuying, _paymentTokenAmount);
        computeRewards(rfs.maker, msg.sender, _amountBuying, rfs.amount0);

rfs.amount0
might be zero ; or is 0 because first we updateRfs (ie. filling RFS in case of test OtcNexusTakeRfsTest::test_takeRfs_failRfsRemoved), then updatingRewards.

I have an error:
[FAIL. Reason: Division or modulo by 0

Reply:
Should be solved with last commit